azure/services/natgateways/spec.go

```diff
@@ -19,88 +19,71 @@ package natgateways
 import (
 	"context"
 
-	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4"
-	"github.com/pkg/errors"
+	asonetworkv1 "github.com/Azure/azure-service-operator/v2/api/network/v1api20220701"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/utils/ptr"
 	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
-	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
-	azureutil "sigs.k8s.io/cluster-api-provider-azure/util/azure"
 )
 
 // NatGatewaySpec defines the specification for a NAT gateway.
 type NatGatewaySpec struct {
 	Name           string
+	Namespace      string
 	ResourceGroup  string
 	SubscriptionID string
 	Location       string
 	NatGatewayIP   infrav1.PublicIPSpec
 	ClusterName    string
 	AdditionalTags infrav1.Tags
+	IsVnetManaged  bool
 }
 
-// ResourceName returns the name of the NAT gateway.
-func (s *NatGatewaySpec) ResourceName() string {
-	return s.Name
-}
-
-// ResourceGroupName returns the name of the resource group.
-func (s *NatGatewaySpec) ResourceGroupName() string {
-	return s.ResourceGroup
-}
-
-// OwnerResourceName is a no-op for NAT gateways.
-func (s *NatGatewaySpec) OwnerResourceName() string {
-	return ""
-}
-
-// Parameters returns the parameters for the NAT gateway.
-func (s *NatGatewaySpec) Parameters(ctx context.Context, existing interface{}) (params interface{}, err error) {
-	if existing != nil {
-		existingNatGateway, ok := existing.(armnetwork.NatGateway)
-		if !ok {
-			return nil, errors.Errorf("%T is not an armnetwork.NatGateway", existing)
-		}
-
-		if hasPublicIP(existingNatGateway, s.NatGatewayIP.Name) {
-			// Skip update for NAT gateway as it exists with expected values
-			return nil, nil
-		}
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *NatGatewaySpec) ResourceRef() *asonetworkv1.NatGateway {
+	return &asonetworkv1.NatGateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
 	}
+}
 
-	natGatewayToCreate := armnetwork.NatGateway{
-		Name:     ptr.To(s.Name),
+// Parameters returns the parameters for the NAT gateway. It implements azure.ASOResourceSpecGetter.
+func (s *NatGatewaySpec) Parameters(ctx context.Context, existingResource *asonetworkv1.NatGateway) (params *asonetworkv1.NatGateway, err error) {
+	// existingResource will be an empty asonetworkv1.NatGateway{} if the aso resource doesn't exist.
+	// if existingResource is not nil, we overwrite the spec with desired values, leaving the status and meta-data as is.
+	// Note that, an aso resource with matching name and namespace can exist but will not get to this point since it will
+	// not be owned by this azure cluster.
```

Comment:
Let me know if this comment is appropriate or if I should reduce it. The aim is to clarify to the reader the reason we "overwrite" the spec.